src/java.base/share/classes/com/sun/crypto/provider/GaloisCounterMode.java

```diff
@@ -198,820 +542,1177 @@ private void checkDataLength(int ... lengths) {
+                        return dst;
+                    }
+                }
+            } else {
+                // buffer types are not the same and can be used as-is
+                return dst;
             }
-            // If blen == 0 and there was not enough to construct a block
-            // from 'buffer' and 'in', then let the below code append 'in' to
-            // the ibuffer.
+
+            // Create a copy
+            ByteBuffer tmp = dst.duplicate();
+            // We can use a heap buffer for internal use, save on alloc cost
+            ByteBuffer bb = ByteBuffer.allocate(dst.remaining());
+            tmp.limit(dst.limit());
+            tmp.position(dst.position());
+            bb.put(tmp);
+            bb.flip();
+            originalDst = dst;
+            return bb;
         }
 
-        // Write any remaining bytes outside the blockSize into ibuffer.
-        int remainder = inLen % blockSize;
-        if (remainder > 0) {
-            if (ibuffer == null) {
-                ibuffer = new ByteArrayOutputStream(inLen % blockSize);
+        /**
+         * Overlap detection for data using byte array.
+         * If an intermediate array is needed, the whole original out array
+         * length is allocated because it's simpler than keep the same offset
+         * and hope the expected output is
+         */
+        byte[] overlapDetection(byte[] in, int inOfs, byte[] out, int outOfs) {
+            if (in == out && inOfs < outOfs) {
+                originalOut = out;
+                originalOutOfs = outOfs;
+                return new byte[out.length];
             }
-            len -= remainder;
-            inLen -= remainder;
-            // remainder offset is based on original buffer length
-            ibuffer.write(in, inOfs + inLen, remainder);
+            return out;
         }
 
-        // Encrypt the remaining blocks inside of 'in'
-        if (inLen > 0) {
-            encryptBlocks(in, inOfs, inLen, out, outOfs);
+        /**
+         * If originalDst is not null, 'dst' is an internal buffer and it's
+         * data will be copied to the original dst buffer
+         */
+        void restoreDst(ByteBuffer dst) {
+            if (originalDst == null) {
+                return;
+            }
+
+            dst.flip();
+            originalDst.put(dst);
         }
 
-        return len;
-    }
+        /**
+         * If originalOut is not null, the 'out' is an internal buffer and it's
+         * data will be copied into original out byte[];
+         */
+        void restoreOut(byte[] out, int len) {
+            if (originalOut == null) {
+                return;
+            }
 
-    void encryptBlocks(byte[] in, int inOfs, int len, byte[] out, int outOfs) {
-        gctrPAndC.update(in, inOfs, len, out, outOfs);
-        processed += len;
-        ghashAllToS.update(out, outOfs, len);
+            System.arraycopy(out, originalOutOfs, originalOut, originalOutOfs,
+                len);
```

Comment:
A engine is a one time use, so setting originalOut to null isn't necessary.  engineDoFinal() sets engine = null.

Reply:
engine is one time use per encryption/decryption. But 'originalOut' is for overlap detection/protection which may be used multiple times during multi-part encryption/decrypion. For each overlapDetection()/restoreOut() pair, the 'originalOut' value should be cleared, otherwise there may be cases where the old value of 'originalOut' gets used?

Reply:
Ok. I see what you are saying.  I had not consider a situation where an update buffer overlapped and doFinal did not.  I'll set originalDst and originalOut to null on their restore methods.

Reply:
Great.